src/search.coffee

```diff
@@ -6,234 +6,252 @@ url = require 'url'
-    opts = _.extend {}, {top: Search.PAGE_SIZE, skip: 0}, options
-
-    # Generate search options for each of the search requests.
-    pairs = _.zip @generateTops(opts.top), @generateSkips(opts.top, opts.skip)
-    requestOptions = _.map pairs, ([top, skip]) ->
-      _.defaults {top, skip}, options
-
-    search = (options, callback) =>
-      @search vertical, options, callback
-
-    async.mapLimit requestOptions, @parallel, search, callback
-
-  search: (vertical, options, callback) ->
-    requestOptions =
-      uri: "#{BING_SEARCH_ENDPOINT}/#{vertical}"
-      qs: _.extend @requestOptions(options),
-        $format: 'json'
-      auth:
-        user: @accountKey
-        pass: @accountKey
+  ###
+  The PAGE_SIZE variable is used for `count` on most API requests. The
+  documentation states that the maximum results is API specific; but, all of
+  the API references just state that "the actual number delivered may be less
+  than requested." From what I've seen, searches typically return ~30  results;
+  however, impartial result sets occur at any `count > 10`. `25` is a good
+  number for avoiding holes in data and duplicate results across pages.
+  ###
+  @PAGE_SIZE = 25
+
+  ###
+  The MAX_RESULTS variable, used for the legacy `top` pagination option, allows
+  for us to return the same default `50` results as v1.0.1 of this library.
+  ###
+  @MAX_RESULTS = 50
+
+  constructor: (@accountKey, @parallelLimit = 10, @useGzip = true) ->
+
+  sanitizeOptions: (options) ->
+    # Default pagination.
+    options = _.defaults options,
+      count: Search.PAGE_SIZE
+      offset: 0
+
+    # Validate market; send as `mkt`.
+    if options.market?
+      options.mkt = options.market if option.market in markets
+      options = _.omit options, 'market'
+
+    options.q = @quote options.q
+    options
+
+  quote: (str) ->
+    str = str.replace '"', '\"' # Escape existing quotes.
+    str.replace /^|$/g, '"'      # Quote entire phrase.
+
+  executeSearch: (options..., callback) ->
+    options = options[0] or {}
+
+    uri = "#{BING_SEARCH_ENDPOINT}/search"
+    if options.endpoint?
+      uri = "#{BING_SEARCH_ENDPOINT}/#{options.endpoint}"
+      options = _.omit options, 'endpoint'
+    qs = @sanitizeOptions options
+
+    req = request {
+      uri, qs
+      headers:
+        'Ocp-Apim-Subscription-Key': @accountKey
       json: true
       gzip: @useGzip
-
-    req = request requestOptions, (err, res, body) ->
+    }, (err, res, body) ->
       unless err or res.statusCode is 200
         err or= new Error "Bad Bing API response #{res.statusCode}"
       return callback err if err
 
-      callback null, body
+      # Filtered searches' data lives within the webPages property.
+      body = body.webPages if body._type is 'SearchResponse'
+      return callback 'Invalid HTTP response body.' if not body?
+
+      # Parse an ID out of result URLs.
+      invalidId = false
+      body.value.forEach (result) ->
```

Comment:
As long as we successfully got a `200` response with a `body` from the API, I think we should be good trusting that `body.value` (or `body.webPages.value`, for web searches) is an array. In the [API reference](https://msdn.microsoft.com/en-us/library/dn760794.aspx#webanswer), they make no comment about `value` ever being anything but an array (the API reference for `value` looks similar for all verticals).